### ISSUE TITLE
OpenStreetMap místo Google Maps

### ISSUE THREAD

**Issue by zbycz** [Clevis/ReKola-Public#49]
Pojďme zde prodiskutovat možnosti použití OpenStreetMap.

Proč vůbec do toho jít? OSM je stejně jako rekola komunitní projekt a oba projekty tak můžou být zajímavé pro stejnou skupinu lidí. Druhá věc je, že pokud se bude dařit dál rekola rozšiřovat, mohli bychom narazit na free limit googlu, což by začalo být dost drahé. 

A nyní jak na to:
- První věc je v čem mapu zobrazovat. Je to možné i zcela bez práce nasadit na stávající JS rozhraní od googlu, což ovšem asi neřeší druhý bod výše. Minimálně do budoucna by bylo lepší používat opensource mapový framework, třeba Leaflet.
- Nad OSM běží i mapa PrahouNaKole, ovšem obecně jsou OSM mapy docela nepřehledné a opět do budoucna by mohlo být zajímavé mít vlastní renderer přesně pro naše potřeby. Nyní by myslím stačil open.mapquest.com
- asi nejlepší mapová vychytávka, která nutně nesouvisí s OSM je možnost generovat Overlay, tedy překryvnou vrstvu transparent PNG. Na ní by mohly být vykresleny - **zóny, stojany, městské kamery a hrobečky** (kde se kola ztatila, či byla poškozena).
  To vše lze generovat samozřejmě dynamicky z databáze. Kola jsou lepší donačítat jako POIky, neboť musí být klikatelná a častěji se mění (nechceme kešovat)
- Ke zvážení ještě je, jestli neslepovat rovnou mapu s overlayem a posílat do mobilu s hlavičkou "kešuj klíďo týden". Asi by to ušetřilo data, ovšem na kešování ty mobily stejně dost kašlou... @enzy ?
- takovou třešničkou je, že OSM má i několik služeb pro Geocoding, tj GPS->ulice... Ale to je opět výhledový problém

**Comment by Piskvor** [comment]
Nápad se mi líbí - jak využití OSM dat, tak i zobrazení pomocí OSM.

Ad 1+2: jak který render - třeba to, co používá _vyhledávač_ PNK, mi připadá poměrně použitelné (mj. zobrazenými povrchy etc.)
ad 3+4: slepení mapy s overlay mi připadá jako logické - při cachování třeba na měsíc to udrží aspoň ty prohlížeče, které na nějaké cachování vůbec hrají (Android 4.0+ už by to měl mít relativně slušné).

Navíc nemusíme data držet v nějaké vlastní db - spousta dat, o které nám jde (stojany etc.) jsou veřejného charakteru (což se pochopitelně netýká věcí rekolům interních, typu "Zóna je vocamcaď pocamcaď a támhle se krade").

**Comment by enzy** [comment]
Já mám hlavní a nejdůležitější obavu a to je výkon na mobilních zařízeních těchto open-source věcí při přidávání dalších a dalších fíčuryn.

**Comment by enzy** [comment]
@zbycz myslím jak zobrazovací knihovnu tak servírování dlaždic

**Comment by filip-jezek** [comment]
Mapa prahou na kole je sice nice, ale strasne neprehledna. Proto sem rad,
ze funguje rekola.cz/map s normalni google mapou..

Imho osm casem proc ne, ale bude to asi chtit trosku predelat render, coz
je zas casove narocny.

Zaverem: proc ne, ale malickatou prioritou, ktera bude rust cim vice se
budeme blizit stropu u gmaps.

**Comment by Piskvor** [comment]
Tak jsem si přes víkend trochu hrál s Tilemillem (a s Nominatimem), a mám jednoduchej proof-of-concept - map tiles i vyhledávání běží nad lokálním extraktem z OSM (resp. nominatimu se ho dotazuje, a map tiles jsou z něj staticky vygenerovaný): http://rekolacz:rekolacz@nominatim.limes.dyndns.org:19401/map/map.html

Jediný, co to tahá "zvenku", je to KML s kolama.

(sedí mi to doma, takže se ty dlaždice budou tahat max. 100 kB/s)

**Comment by Piskvor** [comment]
> Ke zvážení ještě je, jestli neslepovat rovnou mapu s overlayem a posílat do mobilu s hlavičkou "kešuj klíďo týden". Asi by to ušetřilo data, ovšem na kešování ty mobily stejně dost kašlou...

No, i to se dá - když už máme vygenerovaný ty vlastní dlaždice: http://i.piskvor.org/rekola/staticmaplite/index.html
Na kešování většinou nekašlou, když se jim to správně podá (headers a conditional validation); spíš je problém, že při každým zobrazení mapy bude jinde ta poloha z GPS, takže bude jinde marker, takže se stejně bude generovat jiný URL, takže je to dost akademický rozdíl :(

**Comment by Piskvor** [comment]
Trochu jsem si přes víkend hrál s HTTP kešováním a s úpravami renderu.

Vizte všechna ReKola co obrázek (dynamicky generovaný) na pozadí té nově vyrenderované mapy: http://i.piskvor.org/rekola/vsechnakola.php
A tady jako dynamickou mapu: http://i.piskvor.org/rekola/leaflet-example/

Chybí vám v té mapě něco? Přebývá? Co se vám na tom ne/líbí?

**Comment by tsusanka** [comment]
Jen pro pořádek napíšu i sem menší poznámku, že API ma nyní geocoding přes OSM (díky @Piskvor) a v nejbližších dnech se upraví mobilní apka, aby tento geocoding volala. Mapa tam bude stále googlácká, ale geolokace poběží přes rekola API, tudíž OSM.
